src/lib/unlang/compile.c

```diff
@@ -4657,10 +4657,127 @@ static CONF_SECTION *virtual_module_find_cs(CONF_ITEM *ci, rlm_components_t *pco
 	return subcs;
 }
 
+/** Parse per call module env
+ *
+ * Used for config options which must be parsed in the context in which
+ * the module is being called.
+ *
+ * @param[in] single		Module call being compiled.
+ * @param[in] unlang_ctx	Current compilation context.
+ * @param[in] cs		Module config.
+ * @param[in] module_env	to parse.
+ * @return
+ *	- 0 on success;
+ *	- <0 on failure;
+ */
+static int parse_method_env(unlang_module_t *single, unlang_compile_t *unlang_ctx, CONF_SECTION const *cs,
```

Comment:
It might be worth using a slab allocator to hold the `tmpl_t` produced so they occur sequentially in memory.  I think it's a good trade off between difficulty of implementation and at least some speedups.